cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+					goto exit;
+				}
+			}
+		} while (STAILQ_EMPTY(&zinfo->complete_queue));
+
+		zio_cmd = STAILQ_FIRST(&zinfo->complete_queue);
+		STAILQ_REMOVE_HEAD(&zinfo->complete_queue, cmd_link);
+		(void) pthread_mutex_unlock(&zinfo->complete_queue_mutex);
+
+		ASSERT(zio_cmd->conn == fd);
+		ZREPL_LOG("ACK for op:%d with seq-id %ld\n",
+		    zio_cmd->hdr.opcode, zio_cmd->hdr.io_seq);
+
+		rc = uzfs_zvol_socket_write(zio_cmd->conn,
+		    (char *)&zio_cmd->hdr, sizeof (zio_cmd->hdr));
+		if (rc == -1) {
+			ZREPL_ERRLOG("socket write err :%d\n", errno);
+			zio_cmd_free(&zio_cmd, zio_cmd->hdr.opcode);
+			goto exit;
+		}
+
+		switch (zio_cmd->hdr.opcode) {
+			case ZVOL_OPCODE_HANDSHAKE:
+			case ZVOL_OPCODE_WRITE:
+			case ZVOL_OPCODE_SYNC:
+				zinfo->write_req_ack_cnt++;
+				/* Send handsake ack */
+				break;
+			case ZVOL_OPCODE_READ:
+				printf("ACK for op:%d with seq-id %ld\n",
+				    zio_cmd->hdr.opcode, zio_cmd->hdr.io_seq);
+				/* Send data read from disk */
+				rc = uzfs_zvol_socket_write(zio_cmd->conn,
+				    zio_cmd->buf,
+				    (sizeof (char) * zio_cmd->hdr.len));
+				if (rc == -1) {
+					ZREPL_ERRLOG("socket write err :%d\n",
+					    errno);
+					ASSERT(0);
+					goto exit;
+				}
+				zinfo->read_req_ack_cnt++;
+				break;
+
+			default:
+				VERIFY(!"Should be a valid opcode");
+				break;
+		}
+		zio_cmd_free(&zio_cmd, zio_cmd->hdr.opcode);
+	}
+exit:
+	close(fd);
+	while (!STAILQ_EMPTY(&zinfo->complete_queue)) {
+		zio_cmd = STAILQ_FIRST(&zinfo->complete_queue);
+		STAILQ_REMOVE_HEAD(&zinfo->complete_queue, cmd_link);
+		zio_cmd_free(&zio_cmd, zio_cmd->hdr.opcode);
+	}
+	uzfs_zinfo_drop_refcnt(zinfo, false);
+
+	printf("uzfs_zvol_io_ack_sender thread exiting\n");
+	zk_thread_exit();
+}
+
+static void
+uzfs_zrepl_open_log(void)
+{
+	openlog("zrepl", LOG_PID, LOG_LOCAL7);
+}
+
+static void
+uzfs_zrepl_close_log(void)
+{
+	closelog();
+}
+
+static void
+uzfs_zrepl_walk_pool_directory(void)
+{
+	spa_t	*spa = NULL;
+	spa_t	*sp = NULL;
+	int	rc = 0;
+
+	mutex_enter(&spa_namespace_lock);
+	while ((spa = spa_next(spa)) != NULL) {
+		(void) printf(">>>>>>>\t%s\n", spa_name(spa));
+		rc = spa_open(spa_name(spa), &sp, spa);
+		if (rc == 0) {
+			spa_close(spa, spa);
+		}
+	}
+	mutex_exit(&spa_namespace_lock);
+}
+
+/*
+ * Main function for replica.
+ */
+int
+main(void)
+{
+	int 		rc;
+	kthread_t 	*conn_accpt_thrd;
+	kthread_t 	*uzfs_mgmt_thread;
+
+
+	pthread_t slf = pthread_self();
+	snprintf(tinfo, sizeof (tinfo), "m#%d.%d",
+	    (int)(((uint64_t *)slf)[0]), getpid());
+
+	zfs_arc_max = (512 << 20);
+	zfs_arc_min = (256 << 20);
+	printf("zarcmax: %lu zarcmin:%lu\n", zfs_arc_max, zfs_arc_min);
+
+	rc = uzfs_init();
+	uzfs_zrepl_open_log();
+	if (rc != 0) {
+		printf("initialization errored.. %d\n", rc);
+		return (-1);
+	}
+
+	rc = pthread_mutex_init(&zvol_list_mutex, NULL);
+	if (rc != 0) {
+		ZREPL_ERRLOG("zvol_global mutex_init() failed\n");
+		return (-1);
+	}
+
+	uzfs_zrepl_walk_pool_directory();
+	sleep(5);
```

Comment:
hm  🤔  why 5 seconds?

Reply:
Just a random value. This delays is mainly because I want to load/open all datasets incore before someone start establishing connection and pumping data to it. We do little heavy task
loading/opening data set, which may take longer time. I know 5Sec is too much but since this is one time delay which is in boot path so I think it is okay to have some delay(if not 5sec). Please thoughts around ......

I have added this based on UT where I was starting replica and at same time started tgt, where tgt was little faster and it requested LUN/VOL which exist but replica unable to found it because replica was still busy loading/opening it.

Reply:
Thanks for explaining. I think it would be good to add a brief comment explaining the reason here.